shell/platform/windows/win32_flutter_window.h

```diff
@@ -0,0 +1,177 @@
+
+#include <windowsx.h>
+
+#include <string>
+#include <vector>
+
+#include "flutter/shell/platform/common/cpp/client_wrapper/include/flutter/plugin_registrar.h"
+#include "flutter/shell/platform/common/cpp/incoming_message_dispatcher.h"
+#include "flutter/shell/platform/embedder/embedder.h"
+#include "flutter/shell/platform/windows/angle_surface_manager.h"
+#include "flutter/shell/platform/windows/key_event_handler.h"
+#include "flutter/shell/platform/windows/keyboard_hook_handler.h"
+#include "flutter/shell/platform/windows/platform_handler.h"
+#include "flutter/shell/platform/windows/public/flutter_windows.h"
+#include "flutter/shell/platform/windows/text_input_plugin.h"
+#include "flutter/shell/platform/windows/win32_window.h"
+#include "flutter/shell/platform/windows/window_state.h"
+
+namespace flutter {
+
+// A win32 flutter window.  In the future, there will likely be a
+// CoreWindow-based FlutterWindow as well.  At the point may make sense to
+// dependency inject the native window rather than inherit.
+class Win32FlutterWindow : public Win32Window {
+ public:
+  Win32FlutterWindow();
+  Win32FlutterWindow(const char* title,
+                     const int x,
+                     const int y,
+                     const int width,
+                     const int height) noexcept;
+  ~Win32FlutterWindow();
+
+  // Run a Windows message pump that also pumps plugin messages.
+  void FlutterMessageLoop();
+
+  // Virtual method called by base Window when a resize occurs
+  void OnResize(unsigned int width, unsigned int height) override;
+
+  // Virtual method called by base Window when the DPI changes either when a
+  // user drags the window between monitors of differing DPI or when the user
+  // manually changes the scale factor.
+  void OnDpiScale(unsigned int dpi) override;
+
+  // Virtual method called by base window when the pointer moves within the
+  // window bounds.
+  void OnPointerMove(double x, double y) override;
+
+  // Virtual method called by base window when the left mouse button goes down
+  void OnPointerDown(double x, double y) override;
+
+  // Virtual method called by base window when the left mouse button goes from
+  // down to up
+  void OnPointerUp(double x, double y) override;
+
+  // Virtual method called by base window when character input occurs.
+  void OnChar(unsigned int code_point) override;
+
+  // Virtual method called by base window when raw keyboard input occurs.
+  void OnKey(int key, int scancode, int action, int mods) override;
+
+  // Virtual method called by base window when mouse scrollwheel input occurs.
+  void OnScroll(double delta_x, double delta_y) override;
+
+  // Virtual methed called by the base window when the user closes the Windows
+  void OnClose();
+
+  // Configures the window instance with an instance of a running Flutter engine
+  // returning a configured FlutterDesktopWindowControllerRef.
+  FlutterDesktopWindowControllerRef SetState(FLUTTER_API_SYMBOL(FlutterEngine)
```

Comment:
Either the way this is used, or the name, or both, is really confusing to me. It's called SetState, which implies it's setting internal details of this class, but it seems like the primary effect is actually to create and return this entirely new object. (And then, compounding the confusion for me, the caller immediately causes the returned object to own *this* object, inverting what I would have expected from the fact that this is a method on the window).

I don't have a fully thought-through suggestion yet since I haven't been able to spend much time looking at it, but my first thought would be: can you separate this into two separate things: a SetState that does the internal setup, and then a static function or similar that creates the controller ref from a window (and taking ownership of it)?

When I have more time I'll try to diagram this out for myself more and see if I can make a clearer suggestion if that's not viable.

Reply:
I've taken a stab at improving this.  LMK if you need further changes.